### PR TITLE
First version to CR

### PR DESCRIPTION
**Pull request by rlogic**
- Three types of alert
- Copying predefined config from example
- Using helper

### CODE REVIEW

File: .ruby-version [hunk 1]
```diff
@@ -0,0 +1 @@
+ruby-2.4.6
```
Comment (rlogic) [inline .ruby-version]: I haven't seen this file in any other gem. Do we need this?

File: Gemfile [hunk 2]
```diff
@@ -12,3 +12,5 @@ gemspec
 
 # To use a debugger
 # gem 'byebug', group: [:development, :test]
+
+gem 'rubocop', '0.66'
```
Comment (rlogic) [inline Gemfile]: remove from here and add  to difficult-customer.gemspec: `spec.add_development_dependency "rubocop", "0.66"`

File: Gemfile.lock [hunk 3]
```diff
@@ -0,0 +1,137 @@
+PATH
```
Comment (rlogic) [inline Gemfile.lock]: As far as I know there should not be any Gemfile.lock in a gem project

File: app/helpers/difficult_customer_helper.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
+# Rails helper
+module DifficultCustomerHelper
+  def difficult_customer
+    render_partial %w[alert modal page].sample
+    # partials_to_render
+    # render_partial 'alert'
+  end
+
+  def difficult_customer_options
```
Comment (rlogic) [inline app/helpers/difficult_customer_helper.rb]: This should be a private method, right?

File: app/helpers/difficult_customer_helper.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    DifficultCustomer::Config.new.config
+  end
+
+  def fake_classes_array(array_length, string_length: 8)
+    Array.new(array_length) { fake_class(string_length: string_length) }
+  end
+
+  private
+
+  def render_partial(partial)
```
Comment (rlogic) [inline app/helpers/difficult_customer_helper.rb]: kind of a overkill

File: app/helpers/difficult_customer_helper.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    render "difficult_customer/#{partial}"
+  end
+
+  def fake_class(string_length: 8)
+    ''.tap { |s| string_length.times { s << ('a'..'z').to_a.sample } }
```
Comment (rlogic) [inline app/helpers/difficult_customer_helper.rb]: `string_length.times.map { ('a'..'z').to_a.sample }.join`

File: app/helpers/difficult_customer_helper.rb [hunk 7]
```diff
@@ -0,0 +1,28 @@
+# Rails helper
+module DifficultCustomerHelper
+  def difficult_customer
+    render_partial %w[alert modal page].sample
```
Comment (rlogic) [inline app/helpers/difficult_customer_helper.rb]: add a TODO or INFO explaining why it's done this way

File: app/helpers/difficult_customer_helper.rb [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    # partials_to_render
+    # render_partial 'alert'
+  end
+
+  def difficult_customer_options
+    DifficultCustomer::Config.new.config
+  end
+
+  def fake_classes_array(array_length, string_length: 8)
```
Comment (rlogic) [inline app/helpers/difficult_customer_helper.rb]: `random_dom_ids` :), seriously

Comment (rlogic) [reply app/helpers/difficult_customer_helper.rb]: rename array_length to count and string_length to complexity, also change the way how you call the method:

```
def random_dom_ids(count:, complexity: 10)
  Array.new(count) { random_string(complexity: complexity) }
end

private

def random_string(complexity: complexity)
  complexity.times.map { ('a'..'z').to_a.sample }.join
end
```

File: difficult_customer.gemspec [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+                     'MIT_LICENSE',
+                     'Rakefile',
+                     'README.md'
+                    ]
+  s.require_paths = %w[lib]
+
+  s.add_dependency 'bundler', '~> 1.3'
+  s.add_dependency 'rails'
+  s.add_dependency 'rake'
+end
```
Comment (rlogic) [inline difficult_customer.gemspec:31]: as I said use also `add_development_dependency `

File: lib/difficult_customer/config.rb [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+      config_path if File.exist?(config_path)
+      File.join(
+        File.dirname(__FILE__),
+        '..',
+        '..',
+        'example',
+        'difficult_customer.yml'
+      )
+    end
+
```
Comment (rlogic) [inline lib/difficult_customer/config.rb]: remove blank line

File: lib/difficult_customer/engine.rb [hunk 11]
```diff
@@ -0,0 +1,9 @@
+# .
+module DifficultCustomer
+  # A Rails Engine used make the gem running.
+  class Engine < ::Rails::Engine
+
```
Comment (rlogic) [inline lib/difficult_customer/engine.rb:5]: remove

File: lib/difficult_customer/engine.rb [hunk 12]
```diff
@@ -0,0 +1,9 @@
...
+    isolate_namespace DifficultCustomer
+
```
Comment (rlogic) [inline lib/difficult_customer/engine.rb:7]: remove

File: lib/difficult_customer/engine.rb [hunk 13]
```diff
@@ -0,0 +1,9 @@
+# .
+module DifficultCustomer
+  # A Rails Engine used make the gem running.
+  class Engine < ::Rails::Engine
```
Comment (rlogic) [inline lib/difficult_customer/engine.rb:4]: once again because I don't remember: why do we need engine?

File: lib/difficult_customer/helper.rb [hunk 14]
```diff
@@ -0,0 +1 @@
+../app/helpers/difficult_customer_helper.rb
```
Comment (rlogic) [inline lib/difficult_customer/helper.rb:1]: why helpers aren't defined here?

File: lib/generators/difficult_customer/install_generator.rb [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+  module Generators
+    # This generator copies example config into config/.
+    class InstallGenerator < Rails::Generators::Base
+
+      source_root File.expand_path('../../../example', __dir__)
+
+      def copy_config
+        template 'difficult_customer.yml', 'config/difficult_customer.yml'
+      end
+
```
Comment (rlogic) [inline lib/generators/difficult_customer/install_generator.rb:11]: remove

File: lib/generators/difficult_customer/install_generator.rb [hunk 16]
```diff
@@ -0,0 +1,14 @@
+module DifficultCustomer
+  module Generators
+    # This generator copies example config into config/.
+    class InstallGenerator < Rails::Generators::Base
+
```
Comment (rlogic) [inline lib/generators/difficult_customer/install_generator.rb:5]: remove

**Review by rlogic — COMMENTED** [review]
Few things to fix until approved

File: lib/difficult_customer/engine.rb (hunk 13 above)
Comment (mpptx) [reply lib/difficult_customer/engine.rb:4]: Afaik an engine is necessary to make the host app see the files in app/, lib/, etc.

File: lib/difficult_customer/helper.rb (hunk 14 above)
Comment (mpptx) [reply lib/difficult_customer/helper.rb:1]: That allows us to load the gem into app by placing ```= difficult_customer```in the layout file.

**Pull request merged**